### ISSUE TITLE
C implementation of hyperswarm? (IoT / embedded / ESP32)

### ISSUE THREAD

**Issue by Harvie** [hyperswarm/hyperswarm#89]
I think hyperswarm would be great for IoT devices.
Currently you usualy need cloud to control your smart light bulbs and/or domestic heating.
But with hyperswarm it would be possible to control my home/lab directly without middleman.

If we would be able to run hyperswarm directly on wifi-enabled microcontroller (like ESP32), we would be able to handle all communication through p2p network. But usualy it is only feasible to program such tiny devices in C.
Having hyperswarm ported to C would help a lot.

Also there is already LwIP network stack available on embedded devices, which is relatively advanced and should provide all needed APIs for sockets and raw packets.

**Comment by lanmower** [comment]
I second this :) what would it take, moving some more of hyperswarm to native libs? Wouldn't that be really good for the node hyperswarm in the long run too? It doesn't have to be a separate implementation, it can be an optimization for the main package

**Comment by jb0gie** [comment]
I completely agree with your thoughts. The use of hyperswarm for IoT devices could indeed revolutionize the way we interact with our smart devices. The idea of eliminating the need for a cloud-based middleman and enabling direct control over devices through a peer-to-peer network is very appealing.

The potential of running hyperswarm on a wifi-enabled microcontroller like ESP32 opens up a whole new realm of possibilities. It would make the communication process more efficient and direct, which is a significant advantage in the world of IoT.

Porting hyperswarm to C would indeed be beneficial, considering that most microcontrollers are programmed in C due to its efficiency and control over hardware. This would make hyperswarm more accessible and usable in the realm of embedded systems.

The availability of the LwIP network stack on embedded devices is another advantage, as it provides all the necessary APIs for sockets and raw packets. This would make the implementation of hyperswarm on such devices much smoother.

In conclusion, your idea holds a lot of promise and could potentially lead to significant advancements in the field of IoT. It’s an exciting prospect to consider!

**Comment by lanmower** [comment]
For reference sake, here is someone that compiled a realtime PC/MAC library for ESP32

https://github.com/torsoelectronics/link-esp

This is the ableton link library, it's a UDP metronome sync between ocmputers using Kalman filters

**Comment by lanmower** [comment]
There are also some javascript projects for ESP32, that might be a lead if we can somehow make them work with the same native libraries

**Comment by lanmower** [comment]
I feel that some of the work is already done here... I see there's https://www.moddable.com/

There will also be other js implementations, can we rip the beating heart of hyperswarm out of node and put it in one of these other platforms?

**Comment by lanmower** [comment]
![image](https://github.com/holepunchto/hyperswarm/assets/657315/32b42164-5100-4a54-bef7-a94876b13122)

This sounds like exactly what we need for hyperswarm on esp @mafintosh do you think this sounds feasible?

**Comment by lanmower** [comment]
The more I think about this idea the more it makes sense, imagine pairing a little usb hotspot with a esp device, and programming it to rove around, you could have remote controlled drone vehicles and extend their range anywhere where there's internet signal
